Build failure with some compiler flags

```
[  168s] /usr/bin/c++ -Dprime_server_EXPORTS -I/home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534 -I/home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/prime_server -mbranch-protection=standard -O2 -Wall -D_FORTIFY_SOURCE=2 -fstack-protector-strong -funwind-tables -fasynchronous-unwind-tables -fstack-clash-protection -Werror=return-type -flto=auto -g -DNDEBUG -O2 -g -DNDEBUG -fPIC -Wall -Wextra -pedantic -Werror -std=gnu++11 -o CMakeFiles/prime_server.dir/src/zmq_helpers.cpp.o -c /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp: In lambda function:
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp:24:10: error: unused variable 'ret' [-Werror=unused-variable]
[  168s]    24 |     auto ret = zmq_ctx_term(context);
[  168s]       |          ^~~
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp: In lambda function:
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp:40:10: error: unused variable 'ret' [-Werror=unused-variable]
[  168s]    40 |     auto ret = zmq_msg_close(message);
[  168s]       |          ^~~
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp: In lambda function:
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp:56:10: error: unused variable 'ret' [-Werror=unused-variable]
[  168s]    56 |     auto ret = zmq_msg_close(message);
[  168s]       |          ^~~
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp: In lambda function:
[  168s] /home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/src/zmq_helpers.cpp:95:10: error: unused variable 'ret' [-Werror=unused-variable]
[  168s]    95 |     auto ret = zmq_close(socket);
[  168s]       |          ^~~
[  168s] cc1plus: all warnings being treated as errors
[  168s] make[2]: *** [CMakeFiles/prime_server.dir/build.make:137: CMakeFiles/prime_server.dir/src/zmq_helpers.cpp.o] Error 1
[  168s] make[2]: Leaving directory '/home/abuild/rpmbuild/BUILD/prime_server-git.20210301.bfd1534/build'
[  168s] make[2]: *** Waiting for unfinished jobs....
```
Started happening about two weeks ago (right now build fails with the same reason, from master - pulled yesterday).

--- a/.github/workflows/macos.yaml
+++ b/.github/workflows/macos.yaml
@@ -18,7 +18,7 @@ jobs:
         shell: bash
         run: |
           xcrun --show-sdk-path
-          cmake . -DCMAKE_OSX_SYSROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk" && make all -j$(sysctl -n hw.ncpu)
+          cmake . -DCMAKE_BUILD_TYPE=Debug -DCMAKE_OSX_SYSROOT="/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk" && make all -j$(sysctl -n hw.ncpu)
       - name: test
         shell: bash
         run: |

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -7,6 +7,26 @@ INCLUDE(FindPkgConfig)
 # Use a C++11 enabled compiler
 set(CMAKE_CXX_STANDARD 11)
 
+# What type of build
+if(NOT MSVC_IDE) # TODO: May need to be extended for Xcode, CLion, etc.
+  if(NOT CMAKE_BUILD_TYPE)
+    message(STATUS "No build type specified, defaulting to Release")
+    set(CMAKE_BUILD_TYPE Release)
+  endif()
+endif()
+if(CMAKE_BUILD_TYPE MATCHES Debug)
+  message(STATUS "Configuring in debug mode")
+elseif(CMAKE_BUILD_TYPE MATCHES Release)
+  message(STATUS "Configuring in release mode")
+elseif(CMAKE_BUILD_TYPE MATCHES RelWithDebInfo)
+  message(STATUS "Configuring in release mode with debug flags")
+elseif(CMAKE_BUILD_TYPE MATCHES MinRelSize)
+  message(STATUS "Configuring in release mode with minimized size")
+else()
+  set(CMAKE_BUILD_TYPE Release)
+  message(STATUS "Unrecognized build type defaulting to release")
+endif()
+
 # Show all warnings and treat as errors
 if(MSVC)
   add_compile_options(/W4 /WX)

--- a/prime_server/prime_server.hpp
+++ b/prime_server/prime_server.hpp
@@ -133,12 +133,10 @@ public:
   // allows you to favor a certain heartbeat/worker for a given job
   using choose_function_t = std::function<const zmq::message_t*(const std::list<zmq::message_t>&,
                                                                 const std::list<zmq::message_t>&)>;
-  proxy_t(
-      zmq::context_t& context,
-      const std::string& upstream_endpoint,
-      const std::string& downstream_endpoint,
-      const choose_function_t& choose_function =
-          [](const std::list<zmq::message_t>&, const std::list<zmq::message_t>&) { return nullptr; });
+  proxy_t(zmq::context_t& context,
+          const std::string& upstream_endpoint,
+          const std::string& downstream_endpoint,
+          const choose_function_t& choose_function = {});
   virtual ~proxy_t();
   void forward();
 
@@ -174,15 +172,14 @@ public:
       std::function<result_t(const std::list<zmq::message_t>&, void*, interrupt_function_t&)>;
   using cleanup_function_t = std::function<void()>;
 
-  worker_t(
-      zmq::context_t& context,
-      const std::string& upstream_proxy_endpoint,
-      const std::string& downstream_proxy_endpoint,
-      const std::string& result_endpoint,
-      const std::string& interrupt_endpoint,
-      const work_function_t& work_function,
-      const cleanup_function_t& cleanup_function = []() {},
-      const std::string& heart_beat = "");
+  worker_t(zmq::context_t& context,
+           const std::string& upstream_proxy_endpoint,
+           const std::string& downstream_proxy_endpoint,
+           const std::string& result_endpoint,
+           const std::string& interrupt_endpoint,
+           const work_function_t& work_function,
+           const cleanup_function_t& cleanup_function = {},
+           const std::string& heart_beat = "");
   virtual ~worker_t();
   void work();
 

--- a/src/prime_server.cpp
+++ b/src/prime_server.cpp
@@ -432,7 +432,7 @@ void proxy_t::forward() {
         // figure out what worker you want, ignore the request info
         auto info = std::move(messages.front());
         messages.pop_front();
-        const auto* heart_beat = choose_function(fifo, messages);
+        const auto* heart_beat = choose_function ? choose_function(fifo, messages) : nullptr;
         messages.emplace_front(std::move(info));
         // either you didnt want to choose or you sent back garbage
         auto hb_itr = heart_beats.find(heart_beat);
@@ -552,7 +552,8 @@ void worker_t::work() {
 
       // do some cleanup
       try {
-        cleanup_function();
+        if (cleanup_function)
+          cleanup_function();
       } catch (const std::exception& e) {
         logging::ERROR(std::string(__FILE__) + ":" + std::to_string(__LINE__) +
                        " worker_t: " + e.what());

--- a/src/zmq_helpers.cpp
+++ b/src/zmq_helpers.cpp
@@ -1,5 +1,6 @@
 #include "zmq_helpers.hpp"
 #include <arpa/inet.h>
+#include <cerrno>
 #include <ctime>
 #include <czmq.h>
 #include <random>
@@ -22,7 +23,8 @@ context_t::context_t(/*TODO: add options*/) {
   // wrap it in RAII goodness
   ptr.reset(context, [](void* context) {
     auto ret = zmq_ctx_term(context);
-    assert(ret == 0);
+    if (ret != 0)
+      throw std::runtime_error("zmq_ctx_term error: " + std::to_string(errno));
   });
 }
 context_t::operator void*() {
@@ -38,7 +40,8 @@ message_t::message_t(void* data, size_t size, void (*free_function)(void*, void*
   // wrap it in RAII goodness
   ptr.reset(message, [](zmq_msg_t* message) {
     auto ret = zmq_msg_close(message);
-    assert(ret == 0);
+    if (ret != 0)
+      throw std::runtime_error("zmq_msg_close error: " + std::to_string(errno));
     delete message;
   });
 }
@@ -54,7 +57,8 @@ message_t::message_t(size_t size, const void* data) {
   // wrap it in RAII goodness
   ptr.reset(message, [](zmq_msg_t* message) {
     auto ret = zmq_msg_close(message);
-    assert(ret == 0);
+    if (ret != 0)
+      throw std::runtime_error("zmq_msg_close error: " + std::to_string(errno));
     delete message;
   });
 }
@@ -93,7 +97,8 @@ socket_t::socket_t(const context_t& context, int socket_type) : context(context)
   // wrap it in RAII goodness
   ptr.reset(socket, [](void* socket) {
     auto ret = zmq_close(socket);
-    assert(ret == 0);
+    if (ret != 0)
+      throw std::runtime_error("zmq_close error: " + std::to_string(errno));
   });
 }
 // set an option on this socket

